File: api/crs/codegen/lib/crs_solve_trilt/crs_solve_trilt.c
#include "crs_solve_trilt.h"
#include "crs_solve_trilt_types.h"

void crs_solve_trilt(const CRS_Matrix *A, emxArray_real_T *b)
{
  int b_i;
  int i;
  int i1;
  int i2;
  int k;
  int n;
  n = A->row_ptr->size[0] - 1;
  for (i = n; i >= 1; i--) {
    b->data[i - 1] /= A->val->data[A->row_ptr->data[i] - 2];
    b_i = A->row_ptr->data[i - 1];
    i1 = A->row_ptr->data[i] - 2;
    for (k = b_i; k <= i1; k++) {
      i2 = A->col_ind->data[k - 1] - 1;
      b->data[i2] -= A->val->data[k - 1] * b->data[i - 1];
    }
  }
}

void crs_solve_trilt_initialize(void)
{
}

void crs_solve_trilt_terminate(void)
{
}

File: api/crs/crs_solve_trilt.m
function b = crs_solve_trilt(A, b)
% crs_solve_trilt Solves A.'\b, where A is lower triangular
%     b = crs_solve_trilt(A, b)
%  The matrix is assumed to be lower triangular. The right-hand side vector
%  will be overwritten by A.'\b at output
%
% See also crs_solve_tril, crs_solve_triut

%#codegen -args {crs_matrix, coder.typeof(0, [inf,1])}

for i = int32(length(A.row_ptr)) - 1:-1:1
    cend = A.row_ptr(i+1) - 1;
    assert(A.col_ind(cend) == i && A.val(cend) ~= 0);

    b(i) = b(i) / A.val(cend);

    for k = A.row_ptr(i):(cend - 1)
        j = A.col_ind(k);
        % This can be done concurrently
        b(j) = b(j) - A.val(k) * b(i);
    end
end

function test %#ok<DEFNU>
%!test
%! for k=1:100
%!     L = tril(rand(10,10)); b = rand(10, 1);
%!     if condest(L) < 1.e5
%!         A = crs_createFromSparse(sparse(L));
%!         us = crs_solve_trilt(A, b);
%!         assert(norm(L' * us - b) <= 1.e-10);
%!     end
%! end
